cli/commons/src/Microsoft.Kiota.Cli.Commons/IO/ODataPagingService.cs

```diff
@@ -0,0 +1,161 @@
+                    if (!string.IsNullOrWhiteSpace(link)) return new Uri(link);
+                }
+            }
+            catch (JsonException)
+            {
+                // If the response isn't valid JSON, there will be no next link.
+                // TODO: Log warning once logging story is defined
+            }
+        }
+
+        return null;
+    }
+
+    /// <inheritdoc />
+    public async Task<Stream> GetPagedDataAsync(Func<RequestInformation, CancellationToken, Task<Stream>> requestExecutorAsync, PageLinkData pageLinkData, bool fetchAllPages = false, CancellationToken cancellationToken = default)
+    {
+        // Set the page size to 999 if the user asked to fetch all pages and top either isn't specified or is invalid
+        if (fetchAllPages && (!pageLinkData.RequestInformation.QueryParameters.TryGetValue("%24top", out var topVal) || (topVal as int?) < 1))
+        {
+            pageLinkData.RequestInformation.QueryParameters["%24top"] = 999;
+        }
+        var requestInfo = pageLinkData.RequestInformation;
+        Uri? nextLink;
+        Stream? response = null;
+        do
+        {
+            var pageData = await requestExecutorAsync(requestInfo, cancellationToken);
+            if (fetchAllPages)
+            {
+                pageLinkData = new PageLinkData(requestInfo, pageData, pageLinkData.ResponseFormat, pageLinkData.ItemName, pageLinkData.NextLinkName);
+                nextLink = await GetNextPageLinkAsync(pageLinkData, cancellationToken);
+                if (nextLink != null) pageLinkData.RequestInformation.URI = nextLink;
+            }
+            else
+            {
+                nextLink = null;
+            }
+            response = await MergeJsonStreamsAsync(response, pageData, pageLinkData.ItemName, pageLinkData.NextLinkName, cancellationToken);
+        } while (nextLink != null);
+
+        return response ?? Stream.Null;
+    }
+
+    /// <summary>
+    /// Merges 2 streams of JSON on the property defined by <code>itemName</code>. The property should be a JSON array
+    /// </summary>
+    /// <param name="left">The first stream.</param>
+    /// <param name="right">The second stream.</param>
+    /// <param name="itemName">The name of the array property to merge on.</param>
+    /// <param name="nextLinkName">The name of the property containing the next link name.</param>
+    /// <param name="cancellationToken">The cancellation token.</param>
+    internal async Task<Stream?> MergeJsonStreamsAsync(Stream? left, Stream? right, string itemName = "value", string nextLinkName = "nextLink", CancellationToken cancellationToken = default)
+    {
+        if (left?.CanSeek == true) left?.Seek(0, SeekOrigin.Begin);
+        if (right?.CanSeek == true) right?.Seek(0, SeekOrigin.Begin);
+        if (left == null || right == null)
+        {
+            return left ?? right;
+        }
+
+        JsonNode? nodeLeft = null;
+        if (left != null)
```

Comment:
We already know left is not null from line 83

Reply:
No, that line uses a logical OR operator. The code also is a short circuit code that returns if either condition is true. The reason is that if one of the streams is null, then we don't need to run the merge algorithm and can return the non-null stream.

Reply:
That means that the code on line 83 essentially assures us that the left is not null. Otherwise, we would never reach this line it would return. Yes?